Sources/Kafka/KafkaConsumer.swift

```diff
@@ -331,26 +352,89 @@ public final class KafkaConsumer: Sendable, Service {
+                _ = client.eventPoll()
+                try await Task.sleep(for: self.configuration.pollInterval)
+            case .terminatePollLoop:
+                return
+            }
+        }
+    }
+
+    /// Run loop polling Kafka for new consumer messages.
+    private func messageRunLoop() async throws {
         while !Task.isCancelled {
-            let nextAction = self.stateMachine.withLockedValue { $0.nextPollLoopAction() }
+            let nextAction = self.stateMachine.withLockedValue { $0.nextConsumerPollLoopAction() }
             switch nextAction {
             case .pollForAndYieldMessage(let client, let source):
-                let events = client.eventPoll()
-                for event in events {
-                    switch event {
-                    case .consumerMessages(let result):
-                        // We do not support back pressure, we can ignore the yield result
-                        _ = source.yield(result)
-                    default:
-                        break // Ignore
+                // Poll for new consumer message.
+                var result: Result<KafkaConsumerMessage, Error>?
+                do {
+                    if let message = try client.consumerPoll() {
+                        result = .success(message)
+                    }
+                } catch {
+                    result = .failure(error)
+                }
+                if let result {
+                    let yieldResult = source.yield(result)
+                    switch yieldResult {
+                    case .produceMore:
+                        break
+                    case .stopProducing:
+                        self.stateMachine.withLockedValue { $0.stopProducing() }
+                    case .dropped:
+                        break
+                    }
+                } else {
+                    self.stateMachine.withLockedValue { $0.waitForNewMessages() }
+                }
+            case .pollForMessageAndSleep(let client, let source):
+                var result: Result<KafkaConsumerMessage, Error>?
+                do {
+                    if let message = try client.consumerPoll() {
+                        result = .success(message)
                     }
+                } catch {
+                    result = .failure(error)
+                }
+                if let result {
+                    self.stateMachine.withLockedValue { $0.produceMore() }
+                    let yieldResult = source.yield(result)
+                    switch yieldResult {
+                    case .produceMore:
+                        break
+                    case .stopProducing:
+                        self.stateMachine.withLockedValue { $0.stopProducing() }
+                    case .dropped:
+                        break
+                    }
+
+                    self.stateMachine.withLockedValue { $0.newMessagesProduced() }
```

Comment:
I guess It might make sense to continue without sleep after message was produced until we will out of messages again or bump into backpressure.